# 1.Resources of the dataset

https://www.dropbox.com/scl/fo/f4bu0u0qidn9ywvrsd06s/AI6KpLlsKYyjQUmf7tGVJ_I?rlkey=ycpppa0ccukeuj3rd5p0en67l&e=1&dl=0


   # about the dataset
   
       - the loan data is given is a sample data from a larger dataset from a affiliate bank from USA
       - The loan application are filled manually by customers so there must be some missing values

# summary -
     - 1. we are going to convert USD to EURO
     - 2. we will change any text column into numbers based on the info.
     - 3. we will spilt the accounts by months

In [1]:
import numpy as np

In [2]:
np.set_printoptions(suppress = True, linewidth = 100, precision = 2)

## 2.Importing the data

In [3]:
raw_data_np = np.genfromtxt(r"C:\Users\prana\Downloads\S28 - A Loan Data Example with NumPy\loan-data.csv", delimiter = ";", skip_header = 1, autostrip =True)

raw_data_np
#skip_header to remove the column header bcz we need numerical data 
# and autostrip to remove the white spaces 

array([[48010226.  ,         nan,    35000.  , ...,         nan,         nan,     9452.96],
       [57693261.  ,         nan,    30000.  , ...,         nan,         nan,     4679.7 ],
       [59432726.  ,         nan,    15000.  , ...,         nan,         nan,     1969.83],
       ...,
       [50415990.  ,         nan,    10000.  , ...,         nan,         nan,     2185.64],
       [46154151.  ,         nan,         nan, ...,         nan,         nan,     3199.4 ],
       [66055249.  ,         nan,    10000.  , ...,         nan,         nan,      301.9 ]])

## 3.Checking up  Incomplete Data

In [4]:
np.isnan(raw_data_np.sum())

True

In [5]:
np.isnan(raw_data_np).sum()

88005

In [6]:
temp_fill = np.nanmax(raw_data_np)+ 1
temp_fill

68616520.0

In [7]:
filled_dataset = np.where(np.isnan(raw_data_np), temp_fill, raw_data_np)
filled_dataset


array([[48010226.  , 68616520.  ,    35000.  , ..., 68616520.  , 68616520.  ,     9452.96],
       [57693261.  , 68616520.  ,    30000.  , ..., 68616520.  , 68616520.  ,     4679.7 ],
       [59432726.  , 68616520.  ,    15000.  , ..., 68616520.  , 68616520.  ,     1969.83],
       ...,
       [50415990.  , 68616520.  ,    10000.  , ..., 68616520.  , 68616520.  ,     2185.64],
       [46154151.  , 68616520.  , 68616520.  , ..., 68616520.  , 68616520.  ,     3199.4 ],
       [66055249.  , 68616520.  ,    10000.  , ..., 68616520.  , 68616520.  ,      301.9 ]])

In [8]:
np.isnan(temp_fill).sum()
#now all the missiiing data is being filled

0

In [9]:
temp_mean = np.nanmean(raw_data_np, axis = 0)
temp_mean
# we can see few columns are full of NAN values which may be strings

C:\Users\prana\AppData\Local\Temp\ipykernel_26972\966399788.py:1: RuntimeWarning: Mean of empty slice
  temp_mean = np.nanmean(raw_data_np, axis = 0)


array([54015809.19,         nan,    15273.46,         nan,    15311.04,         nan,       16.62,
            440.92,         nan,         nan,         nan,         nan,         nan,     3143.85])

In [10]:
temp_stats =  np.array([np.nanmin(raw_data_np, axis = 0), temp_mean, np.nanmax(raw_data_np, axis = 0)])
temp_stats

C:\Users\prana\AppData\Local\Temp\ipykernel_26972\3645683317.py:1: RuntimeWarning: All-NaN slice encountered
  temp_stats =  np.array([np.nanmin(raw_data_np, axis = 0), temp_mean, np.nanmax(raw_data_np, axis = 0)])


array([[  373332.  ,         nan,     1000.  ,         nan,     1000.  ,         nan,        6.  ,
              31.42,         nan,         nan,         nan,         nan,         nan,        0.  ],
       [54015809.19,         nan,    15273.46,         nan,    15311.04,         nan,       16.62,
             440.92,         nan,         nan,         nan,         nan,         nan,     3143.85],
       [68616519.  ,         nan,    35000.  ,         nan,    35000.  ,         nan,       28.99,
            1372.97,         nan,         nan,         nan,         nan,         nan,    41913.62]])

## 4. a)  Splitting the Dataset
 to see which column is str and which is numeric

In [11]:
#for string column
str_column = np.argwhere(np.isnan(temp_mean)).squeeze()
str_column

array([ 1,  3,  5,  8,  9, 10, 11, 12], dtype=int64)

In [12]:
#for numerical column 
num_column = np.argwhere(np.isnan(temp_mean)== False).squeeze()
num_column 

array([ 0,  2,  4,  6,  7, 13], dtype=int64)

## b) Reimporting the Dataset 

#### for string dataset

In [13]:
loan_data_strings = np.genfromtxt(r"C:\Users\prana\Downloads\S28 - A Loan Data Example with NumPy\loan-data.csv", 
                            delimiter = ";", 
                            skip_header = 1,
                            autostrip =True,
                            usecols = str_column,
                            dtype =str)
loan_data_strings

array([['May-15', 'Current', '36 months', ..., 'Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=48010226', 'CA'],
       ['', 'Current', '36 months', ..., 'Source Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=57693261', 'NY'],
       ['Sep-15', 'Current', '36 months', ..., 'Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=59432726', 'PA'],
       ...,
       ['Jun-15', 'Current', '36 months', ..., 'Source Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=50415990', 'CA'],
       ['Apr-15', 'Current', '36 months', ..., 'Source Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=46154151', 'OH'],
       ['Dec-15', 'Current', '36 months', ..., '',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=66055249', 'IL']],
      dtype='<U69')

#### for numeric dataset

In [14]:
loan_data_numeric = np.genfromtxt(r"C:\Users\prana\Downloads\S28 - A Loan Data Example with NumPy\loan-data.csv", 
                            delimiter = ";", 
                            skip_header = 1,
                            autostrip =True,
                            usecols = num_column,
                            filling_values = temp_fill)
loan_data_numeric

array([[48010226.  ,    35000.  ,    35000.  ,       13.33,     1184.86,     9452.96],
       [57693261.  ,    30000.  ,    30000.  , 68616520.  ,      938.57,     4679.7 ],
       [59432726.  ,    15000.  ,    15000.  , 68616520.  ,      494.86,     1969.83],
       ...,
       [50415990.  ,    10000.  ,    10000.  , 68616520.  , 68616520.  ,     2185.64],
       [46154151.  , 68616520.  ,    10000.  ,       16.55,      354.3 ,     3199.4 ],
       [66055249.  ,    10000.  ,    10000.  , 68616520.  ,      309.97,      301.9 ]])

### The name of the columns

In [15]:
header_full = np.genfromtxt(r"C:\Users\prana\Downloads\S28 - A Loan Data Example with NumPy\loan-data.csv", 
                            delimiter = ";", 
                            skip_footer = raw_data_np.shape[0],
                            autostrip =True,
                            
                            dtype =str)
header_full

# skip_footer with this condition show only header of  the column

array(['id', 'issue_d', 'loan_amnt', 'loan_status', 'funded_amnt', 'term', 'int_rate',
       'installment', 'grade', 'sub_grade', 'verification_status', 'url', 'addr_state',
       'total_pymnt'], dtype='<U19')

In [16]:
header_strings = header_full[str_column]   
header_strings

array(['issue_d', 'loan_status', 'term', 'grade', 'sub_grade', 'verification_status', 'url',
       'addr_state'], dtype='<U19')

In [17]:
loan_data_strings

array([['May-15', 'Current', '36 months', ..., 'Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=48010226', 'CA'],
       ['', 'Current', '36 months', ..., 'Source Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=57693261', 'NY'],
       ['Sep-15', 'Current', '36 months', ..., 'Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=59432726', 'PA'],
       ...,
       ['Jun-15', 'Current', '36 months', ..., 'Source Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=50415990', 'CA'],
       ['Apr-15', 'Current', '36 months', ..., 'Source Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=46154151', 'OH'],
       ['Dec-15', 'Current', '36 months', ..., '',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=66055249', 'IL']],
      dtype='<U69')

In [18]:
header_numeric = header_full[num_column]
header_numeric

array(['id', 'loan_amnt', 'funded_amnt', 'int_rate', 'installment', 'total_pymnt'], dtype='<U19')


# 5- Creating Checkpoints

checkpoints : places through our code where we store a copy of our dataset bcz we want to avoid losing a progress

In [19]:
def checkpoints(file_name, checkpoint_header, checkpoint_data):
    np.savez(file_name, header = checkpoint_header, data = checkpoint_data)
    checkpoint_variable = np.load(file_name+ ".npz")
    return(checkpoint_variable)

In [20]:
checkpoint_test = checkpoints("checkpoint-test", header_strings, loan_data_strings) 

In [21]:
#checking checkpoint
checkpoint_test["header"]

array(['issue_d', 'loan_status', 'term', 'grade', 'sub_grade', 'verification_status', 'url',
       'addr_state'], dtype='<U19')

In [22]:
checkpoint_test["data"]

array([['May-15', 'Current', '36 months', ..., 'Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=48010226', 'CA'],
       ['', 'Current', '36 months', ..., 'Source Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=57693261', 'NY'],
       ['Sep-15', 'Current', '36 months', ..., 'Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=59432726', 'PA'],
       ...,
       ['Jun-15', 'Current', '36 months', ..., 'Source Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=50415990', 'CA'],
       ['Apr-15', 'Current', '36 months', ..., 'Source Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=46154151', 'OH'],
       ['Dec-15', 'Current', '36 months', ..., '',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=66055249', 'IL']],
      dtype='<U69')


# 6- Manipulating The Strings Column

In [23]:
header_strings

array(['issue_d', 'loan_status', 'term', 'grade', 'sub_grade', 'verification_status', 'url',
       'addr_state'], dtype='<U19')

In [24]:
header_strings[0] ="issue_date"
#since issue_d name was not descrptive

In [25]:
header_strings

array(['issue_date', 'loan_status', 'term', 'grade', 'sub_grade', 'verification_status', 'url',
       'addr_state'], dtype='<U19')

In [26]:
loan_data_strings

array([['May-15', 'Current', '36 months', ..., 'Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=48010226', 'CA'],
       ['', 'Current', '36 months', ..., 'Source Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=57693261', 'NY'],
       ['Sep-15', 'Current', '36 months', ..., 'Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=59432726', 'PA'],
       ...,
       ['Jun-15', 'Current', '36 months', ..., 'Source Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=50415990', 'CA'],
       ['Apr-15', 'Current', '36 months', ..., 'Source Verified',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=46154151', 'OH'],
       ['Dec-15', 'Current', '36 months', ..., '',
        'https://www.lendingclub.com/browse/loanDetail.action?loan_id=66055249', 'IL']],
      dtype='<U69')

 alll the records from are from 
 2015 so we can remove(strip) 2015

In [27]:
np.unique(loan_data_strings[:,0])

array(['', 'Apr-15', 'Aug-15', 'Dec-15', 'Feb-15', 'Jan-15', 'Jul-15', 'Jun-15', 'Mar-15',
       'May-15', 'Nov-15', 'Oct-15', 'Sep-15'], dtype='<U69')

In [28]:
loan_data_strings[:,0] = np.chararray.strip(loan_data_strings[:,0], "-15")

In [29]:
np.unique(loan_data_strings[:,0])

array(['', 'Apr', 'Aug', 'Dec', 'Feb', 'Jan', 'Jul', 'Jun', 'Mar', 'May', 'Nov', 'Oct', 'Sep'],
      dtype='<U69')

here we can see the values of issue_date is in strings so we gonna change it with month number using **for loop**

In [30]:
months = np.array(['', 'Apr', 'Aug', 'Dec', 'Feb', 'Jan', 'Jul', 'Jun', 'Mar', 'May', 'Nov', 'Oct', 'Sep'])

In [31]:
for i in range(13):
    loan_data_strings[:,0] = np.where(loan_data_strings[:,0] == months[i],i,loan_data_strings[:,0])

In [32]:
np.unique(loan_data_strings[:,0])

array(['0', '1', '10', '11', '12', '2', '3', '4', '5', '6', '7', '8', '9'], dtype='<U69')

# 7) Manipulating Text Data: Loan Status and Term

## Loan Status

In [33]:
header_strings

array(['issue_date', 'loan_status', 'term', 'grade', 'sub_grade', 'verification_status', 'url',
       'addr_state'], dtype='<U19')

In [34]:
loan_data_strings[:,1]
#we seeing only "current " values but there must be other values too 

array(['Current', 'Current', 'Current', ..., 'Current', 'Current', 'Current'], dtype='<U69')

In [35]:
np.unique(loan_data_strings[:,1])

array(['', 'Charged Off', 'Current', 'Default', 'Fully Paid', 'In Grace Period', 'Issued',
       'Late (16-30 days)', 'Late (31-120 days)'], dtype='<U69')

In [36]:
np.unique(loan_data_strings[:,1]).size
#there are 9 unique values

9

### Regression care if candidates are in stable financial condition

#### so we need to split all possible values in groups (good or bad)

##### **Current, fully Paid, In Grace Period,  Issued and Late 16-30 day** are in good group 
##### Charged off , Default and Late 31-120 days are in bad group 

#### we give numeric values to good and bad
##### good = 1
##### bad = 0

In [37]:
status_bad = np.array(['','Charged Off','Default','Late (31-120 days)'])

In [38]:
loan_data_strings[:,1]= np.where(np.isin(loan_data_strings[:,1], status_bad),0,1)

In [39]:
(loan_data_strings[:,1])

array(['1', '1', '1', ..., '1', '1', '1'], dtype='<U69')

In [40]:
np.unique(loan_data_strings[:,1])

array(['0', '1'], dtype='<U69')

## term column

In [41]:
header_strings

array(['issue_date', 'loan_status', 'term', 'grade', 'sub_grade', 'verification_status', 'url',
       'addr_state'], dtype='<U19')

In [42]:
loan_data_strings[:,2]

array(['36 months', '36 months', '36 months', ..., '36 months', '36 months', '36 months'],
      dtype='<U69')

In [43]:
np.unique(loan_data_strings[:,2])

array(['', '36 months', '60 months'], dtype='<U69')

**striping months from term column**

In [44]:
loan_data_strings[:,2] = np.chararray.strip(loan_data_strings[:,2]," months")

In [45]:
loan_data_strings[:,2]

array(['36', '36', '36', ..., '36', '36', '36'], dtype='<U69')

In [46]:
np.unique(loan_data_strings[:,2])

array(['', '36', '60'], dtype='<U69')

**we can assume 60 months is a worst month and 36 months is good month**

In [47]:
header_strings[2] = "term_months"

In [48]:
header_strings

array(['issue_date', 'loan_status', 'term_months', 'grade', 'sub_grade', 'verification_status',
       'url', 'addr_state'], dtype='<U19')

In [49]:
loan_data_strings[:,2] 

array(['36', '36', '36', ..., '36', '36', '36'], dtype='<U69')

**coverting empty months to 60 months**

In [50]:
loan_data_strings[:,2] = np.where(loan_data_strings[:,2]=='', 60, loan_data_strings[:,2])

In [51]:
np.unique(loan_data_strings[:,2])

array(['36', '60'], dtype='<U69')

## Grade and Subgrade

###### grade

In [52]:
header_strings 

array(['issue_date', 'loan_status', 'term_months', 'grade', 'sub_grade', 'verification_status',
       'url', 'addr_state'], dtype='<U19')

In [53]:
loan_data_strings[:,3]

array(['C', 'A', 'B', ..., 'A', 'D', 'A'], dtype='<U69')

In [54]:
np.unique(loan_data_strings[:,3])

array(['', 'A', 'B', 'C', 'D', 'E', 'F', 'G'], dtype='<U69')

###### sub_grade

In [55]:
loan_data_strings[:,4]

array(['C3', 'A5', 'B5', ..., 'A5', 'D2', 'A4'], dtype='<U69')

In [56]:
np.unique(loan_data_strings[:,4])

array(['', 'A1', 'A2', 'A3', 'A4', 'A5', 'B1', 'B2', 'B3', 'B4', 'B5', 'C1', 'C2', 'C3', 'C4',
       'C5', 'D1', 'D2', 'D3', 'D4', 'D5', 'E1', 'E2', 'E3', 'E4', 'E5', 'F1', 'F2', 'F3', 'F4',
       'F5', 'G1', 'G2', 'G3', 'G4', 'G5'], dtype='<U69')

checking how many times here is missing value

In [57]:
np.unique(loan_data_strings[:,4], return_counts = True)

(array(['', 'A1', 'A2', 'A3', 'A4', 'A5', 'B1', 'B2', 'B3', 'B4', 'B5', 'C1', 'C2', 'C3', 'C4',
        'C5', 'D1', 'D2', 'D3', 'D4', 'D5', 'E1', 'E2', 'E3', 'E4', 'E5', 'F1', 'F2', 'F3', 'F4',
        'F5', 'G1', 'G2', 'G3', 'G4', 'G5'], dtype='<U69'),
 array([514, 285, 278, 239, 323, 502, 509, 517, 530, 553, 494, 629, 567, 586, 564, 423, 391, 267,
        250, 255, 223, 235, 162, 171, 139, 114,  94,  52,  34,  43,  16,  19,  10,   3,   7,   2],
       dtype=int64))

**we see only 514 missing data out of 10000 rows so gonna give them the least grade**|

### Filling subgrades


In [58]:
loan_data_strings[:,4] = np.where(loan_data_strings[:,4]== '', "H1",
                                 loan_data_strings[:,4])

In [59]:
np.unique(loan_data_strings[:,4])

array(['A1', 'A2', 'A3', 'A4', 'A5', 'B1', 'B2', 'B3', 'B4', 'B5', 'C1', 'C2', 'C3', 'C4', 'C5',
       'D1', 'D2', 'D3', 'D4', 'D5', 'E1', 'E2', 'E3', 'E4', 'E5', 'F1', 'F2', 'F3', 'F4', 'F5',
       'G1', 'G2', 'G3', 'G4', 'G5', 'H1'], dtype='<U69')

###### we got sub_grade so we dont need grade anymore

### Removing Grade

In [60]:
#deleting grade coulumn
loan_data_strings = np.delete(loan_data_strings, 3, axis =1)

In [61]:
#deleting grade header
header_strings = np.delete(header_strings, 3)

In [62]:
header_strings

array(['issue_date', 'loan_status', 'term_months', 'sub_grade', 'verification_status', 'url',
       'addr_state'], dtype='<U19')

In [63]:
header_strings[3]

'sub_grade'

### converting subgrade into integers

In [64]:
np.unique(loan_data_strings[:,3])

array(['A1', 'A2', 'A3', 'A4', 'A5', 'B1', 'B2', 'B3', 'B4', 'B5', 'C1', 'C2', 'C3', 'C4', 'C5',
       'D1', 'D2', 'D3', 'D4', 'D5', 'E1', 'E2', 'E3', 'E4', 'E5', 'F1', 'F2', 'F3', 'F4', 'F5',
       'G1', 'G2', 'G3', 'G4', 'G5', 'H1'], dtype='<U69')

In [65]:
keys = list(np.unique(loan_data_strings[:,3]))
values = list(range(1, len(keys) + 1))
dict_sub_grade = dict(zip(keys, values))


In [66]:
dict_sub_grade

{'A1': 1,
 'A2': 2,
 'A3': 3,
 'A4': 4,
 'A5': 5,
 'B1': 6,
 'B2': 7,
 'B3': 8,
 'B4': 9,
 'B5': 10,
 'C1': 11,
 'C2': 12,
 'C3': 13,
 'C4': 14,
 'C5': 15,
 'D1': 16,
 'D2': 17,
 'D3': 18,
 'D4': 19,
 'D5': 20,
 'E1': 21,
 'E2': 22,
 'E3': 23,
 'E4': 24,
 'E5': 25,
 'F1': 26,
 'F2': 27,
 'F3': 28,
 'F4': 29,
 'F5': 30,
 'G1': 31,
 'G2': 32,
 'G3': 33,
 'G4': 34,
 'G5': 35,
 'H1': 36}

##### replacing associated values with it

In [67]:
for i in np.unique(loan_data_strings[:,3]):
    loan_data_strings[:,3]= np.where(loan_data_strings[:,3]== i ,
                                    dict_sub_grade[i],
                                    loan_data_strings[:,3])

In [68]:
np.unique(loan_data_strings[:,3])

array(['1', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '2', '20', '21', '22',
       '23', '24', '25', '26', '27', '28', '29', '3', '30', '31', '32', '33', '34', '35', '36',
       '4', '5', '6', '7', '8', '9'], dtype='<U69')

# 8) Manipulating Text Data : Verification Status & URL

In [69]:
header_strings

array(['issue_date', 'loan_status', 'term_months', 'sub_grade', 'verification_status', 'url',
       'addr_state'], dtype='<U19')

In [70]:
np.unique(loan_data_strings[:,4])

array(['', 'Not Verified', 'Source Verified', 'Verified'], dtype='<U69')

source verified = loan verified which includes investors backing

so we can assume that **source verified and verified as good** and 
 
 
**not verified and '' as bad**

so **we convert good as 1 and bad as 0**


In [71]:
loan_data_strings[:,4] = np.where((loan_data_strings[:,4]=='') | (loan_data_strings[:,4]== 'Not Verified'),0,1)

In [72]:
np.unique(loan_data_strings[:,4])

array(['0', '1'], dtype='<U69')

## URL

In [73]:
loan_data_strings[:,5]

array(['https://www.lendingclub.com/browse/loanDetail.action?loan_id=48010226',
       'https://www.lendingclub.com/browse/loanDetail.action?loan_id=57693261',
       'https://www.lendingclub.com/browse/loanDetail.action?loan_id=59432726', ...,
       'https://www.lendingclub.com/browse/loanDetail.action?loan_id=50415990',
       'https://www.lendingclub.com/browse/loanDetail.action?loan_id=46154151',
       'https://www.lendingclub.com/browse/loanDetail.action?loan_id=66055249'], dtype='<U69')

In [74]:
loan_data_strings[:,5] = np.chararray.strip(loan_data_strings[:,5],"https://www.lendingclub.com/browse/loanDetail.action?loan_id=" )

In [75]:
a = loan_data_strings[:,5].astype(dtype= int)

In [76]:
b = loan_data_numeric[:,0].astype(dtype = int)

In [77]:
np.array_equal(a,b)

True

we can see both have same values so we remove this column

In [78]:
loan_data_strings = np.delete(loan_data_strings, 5, axis = 1)
header_strings = np.delete(header_strings, 5)

In [79]:
loan_data_strings

array([['9', '1', '36', '13', '1', 'CA'],
       ['0', '1', '36', '5', '1', 'NY'],
       ['12', '1', '36', '10', '1', 'PA'],
       ...,
       ['7', '1', '36', '5', '1', 'CA'],
       ['1', '1', '36', '17', '1', 'OH'],
       ['3', '1', '36', '4', '0', 'IL']], dtype='<U69')

# 9)Manipulating Text Data: State Address

In [80]:
header_strings

array(['issue_date', 'loan_status', 'term_months', 'sub_grade', 'verification_status',
       'addr_state'], dtype='<U19')

In [81]:
header_strings[5]

'addr_state'

In [82]:
loan_data_strings[:,5]

array(['CA', 'NY', 'PA', ..., 'CA', 'OH', 'IL'], dtype='<U69')

In [83]:
np.unique(loan_data_strings[:,5])

array(['', 'AK', 'AL', 'AR', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL', 'GA', 'HI', 'IL', 'IN',
       'KS', 'KY', 'LA', 'MA', 'MD', 'ME', 'MI', 'MN', 'MO', 'MS', 'MT', 'NC', 'ND', 'NE', 'NH',
       'NJ', 'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA', 'RI', 'SC', 'SD', 'TN', 'TX', 'UT', 'VA',
       'VT', 'WA', 'WI', 'WV', 'WY'], dtype='<U69')

In [84]:
np.unique(loan_data_strings[:,5]).size

50

In [85]:
np.unique(loan_data_strings[:,5], return_counts = True)

(array(['', 'AK', 'AL', 'AR', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL', 'GA', 'HI', 'IL', 'IN',
        'KS', 'KY', 'LA', 'MA', 'MD', 'ME', 'MI', 'MN', 'MO', 'MS', 'MT', 'NC', 'ND', 'NE', 'NH',
        'NJ', 'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA', 'RI', 'SC', 'SD', 'TN', 'TX', 'UT', 'VA',
        'VT', 'WA', 'WI', 'WV', 'WY'], dtype='<U69'),
 array([ 500,   26,  119,   74,  220, 1336,  201,  143,   27,   27,  690,  321,   44,  389,  152,
          84,   84,  116,  210,  222,   10,  267,  156,  160,   61,   28,  261,   16,   25,   58,
         341,   57,  130,  777,  312,   83,  108,  320,   40,  107,   24,  143,  758,   74,  242,
          17,  216,  148,   49,   27], dtype=int64))

assigning variables to both

In [86]:
states_names, states_count =np.unique(loan_data_strings[:,5], return_counts = True)
#sorting
states_count_sorted = np.argsort(-states_count)
states_names[states_count_sorted], states_count[states_count_sorted]

(array(['CA', 'NY', 'TX', 'FL', '', 'IL', 'NJ', 'GA', 'PA', 'OH', 'MI', 'NC', 'VA', 'MD', 'AZ',
        'WA', 'MA', 'CO', 'MO', 'MN', 'IN', 'WI', 'CT', 'TN', 'NV', 'AL', 'LA', 'OR', 'SC', 'KY',
        'KS', 'OK', 'UT', 'AR', 'MS', 'NH', 'NM', 'WV', 'HI', 'RI', 'MT', 'DE', 'DC', 'WY', 'AK',
        'NE', 'SD', 'VT', 'ND', 'ME'], dtype='<U69'),
 array([1336,  777,  758,  690,  500,  389,  341,  321,  320,  312,  267,  261,  242,  222,  220,
         216,  210,  201,  160,  156,  152,  148,  143,  143,  130,  119,  116,  108,  107,   84,
          84,   83,   74,   74,   61,   58,   57,   49,   44,   40,   28,   27,   27,   27,   26,
          25,   24,   17,   16,   10], dtype=int64))

### Filling null values

In [87]:
loan_data_strings[:,5]= np.where(loan_data_strings[:,5]=='',0,loan_data_strings[:,5] )
#checking
np.unique(loan_data_strings[:,5])

array(['0', 'AK', 'AL', 'AR', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL', 'GA', 'HI', 'IL', 'IN',
       'KS', 'KY', 'LA', 'MA', 'MD', 'ME', 'MI', 'MN', 'MO', 'MS', 'MT', 'NC', 'ND', 'NE', 'NH',
       'NJ', 'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA', 'RI', 'SC', 'SD', 'TN', 'TX', 'UT', 'VA',
       'VT', 'WA', 'WI', 'WV', 'WY'], dtype='<U69')

in few states there are very few data to examine indivually so gonna group them in regions

https://www2.census.gov/geo/pdfs/maps-data/maps/loan_data_strings[:,5]reference/us_regdiv.pdf

In [88]:
states_west = np.array(['WA', 'OR','CA','NV','ID','MT', 'WY','UT','CO', 'AZ','NM','HI','AK'])
states_south = np.array(['TX','OK','AR','LA','MS','AL','TN','KY','FL','GA','SC','NC','VA','WV','MD','DE','DC'])
states_midwest = np.array(['ND','SD','NE','KS','MN','IA','MO','WI','IL','IN','MI','OH'])
states_east = np.array(['PA','NY','NJ','CT','MA','VT','NH','ME','RI'])

assigning numbers to each region accordingly for numerical ananlysis

In [89]:
loan_data_strings[:,5] = np.where(np.isin(loan_data_strings[:,5], states_west), 1, loan_data_strings[:,5])
loan_data_strings[:,5] = np.where(np.isin(loan_data_strings[:,5], states_south), 2, loan_data_strings[:,5])
loan_data_strings[:,5] = np.where(np.isin(loan_data_strings[:,5], states_midwest), 3, loan_data_strings[:,5])
loan_data_strings[:,5] = np.where(np.isin(loan_data_strings[:,5], states_east), 4, loan_data_strings[:,5])

In [90]:
np.unique(loan_data_strings[:,5])

array(['0', '1', '2', '3', '4'], dtype='<U69')

# 10) 213. Manipulating Text Data: Converting Strings and Creating a Checkpoint

## converting to numbers

In [91]:
loan_data_strings

array([['9', '1', '36', '13', '1', '1'],
       ['0', '1', '36', '5', '1', '4'],
       ['12', '1', '36', '10', '1', '4'],
       ...,
       ['7', '1', '36', '5', '1', '1'],
       ['1', '1', '36', '17', '1', '3'],
       ['3', '1', '36', '4', '0', '3']], dtype='<U69')

In [92]:
loan_data_strings = np.int32(loan_data_strings)
loan_data_strings

array([[ 9,  1, 36, 13,  1,  1],
       [ 0,  1, 36,  5,  1,  4],
       [12,  1, 36, 10,  1,  4],
       ...,
       [ 7,  1, 36,  5,  1,  1],
       [ 1,  1, 36, 17,  1,  3],
       [ 3,  1, 36,  4,  0,  3]])

### Making Checkpoint

In [93]:
checkpoint_strings = checkpoints("Checkpoints-Strings", header_strings, loan_data_strings)

In [94]:
checkpoint_strings["header"]

array(['issue_date', 'loan_status', 'term_months', 'sub_grade', 'verification_status',
       'addr_state'], dtype='<U19')

In [95]:
checkpoint_strings["data"]

array([[ 9,  1, 36, 13,  1,  1],
       [ 0,  1, 36,  5,  1,  4],
       [12,  1, 36, 10,  1,  4],
       ...,
       [ 7,  1, 36,  5,  1,  1],
       [ 1,  1, 36, 17,  1,  3],
       [ 3,  1, 36,  4,  0,  3]])

### checkking where equal to or not



In [96]:
loan_data_strings


array([[ 9,  1, 36, 13,  1,  1],
       [ 0,  1, 36,  5,  1,  4],
       [12,  1, 36, 10,  1,  4],
       ...,
       [ 7,  1, 36,  5,  1,  1],
       [ 1,  1, 36, 17,  1,  3],
       [ 3,  1, 36,  4,  0,  3]])

seems both look similar so going to verify

In [97]:
np.array_equal(checkpoint_strings["data"],loan_data_strings)

True

# 11) Manipulating Numeric Data : Substituting Filler Values

## Manipulating Numeric Columns

In [98]:
loan_data_numeric

array([[48010226.  ,    35000.  ,    35000.  ,       13.33,     1184.86,     9452.96],
       [57693261.  ,    30000.  ,    30000.  , 68616520.  ,      938.57,     4679.7 ],
       [59432726.  ,    15000.  ,    15000.  , 68616520.  ,      494.86,     1969.83],
       ...,
       [50415990.  ,    10000.  ,    10000.  , 68616520.  , 68616520.  ,     2185.64],
       [46154151.  , 68616520.  ,    10000.  ,       16.55,      354.3 ,     3199.4 ],
       [66055249.  ,    10000.  ,    10000.  , 68616520.  ,      309.97,      301.9 ]])

In [99]:
np.isnan(loan_data_numeric).sum()
#checking missing values

0

## Substituting "Filler" Values

In [100]:
header_numeric

array(['id', 'loan_amnt', 'funded_amnt', 'int_rate', 'installment', 'total_pymnt'], dtype='<U19')

In [101]:
temp_fill

68616520.0

checking how many times we filled missing values in the first column and 500 in second column

### ID

In [102]:
np.isin(loan_data_numeric[:,0], temp_fill).sum()

0

In [103]:
np.isin(loan_data_numeric[:,1], temp_fill).sum()

500

In [104]:
header_numeric

array(['id', 'loan_amnt', 'funded_amnt', 'int_rate', 'installment', 'total_pymnt'], dtype='<U19')

### Temporary Stats

In [105]:
temp_stats

array([[  373332.  ,         nan,     1000.  ,         nan,     1000.  ,         nan,        6.  ,
              31.42,         nan,         nan,         nan,         nan,         nan,        0.  ],
       [54015809.19,         nan,    15273.46,         nan,    15311.04,         nan,       16.62,
             440.92,         nan,         nan,         nan,         nan,         nan,     3143.85],
       [68616519.  ,         nan,    35000.  ,         nan,    35000.  ,         nan,       28.99,
            1372.97,         nan,         nan,         nan,         nan,         nan,    41913.62]])

In [106]:
temp_stats[:, num_column]
#first row is temp min
#second row is temp mean
#third row is temp max

array([[  373332.  ,     1000.  ,     1000.  ,        6.  ,       31.42,        0.  ],
       [54015809.19,    15273.46,    15311.04,       16.62,      440.92,     3143.85],
       [68616519.  ,    35000.  ,    35000.  ,       28.99,     1372.97,    41913.62]])

## Funded amount
its values should be minimum
because


    -Concentration Risk: This is the risk that a large portion of a lender's portfolio is concentrated in a particular industry, geographic region, or borrower.It increases the vulnerability to economic downturns or industry-specific shocks.  


    -Default Risk: A larger loan amount implies a higher potential loss in case of default.
   
    -Regulatory Requirements: Many financial institutions are subject to regulatory capital requirements that increase with exposure to a single borrower.

In [107]:
loan_data_numeric[:,2]

array([35000., 30000., 15000., ..., 10000., 10000., 10000.])

In [108]:
loan_data_numeric[:,2] = np.where(loan_data_numeric[:,2]== temp_fill,
                                 temp_stats[0, num_column[2]],
                                 loan_data_numeric[:,2])
loan_data_numeric[:,2]

array([35000., 30000., 15000., ..., 10000., 10000., 10000.])

### Assigning Max Temporary values to 
**Loaned Amount, Interest Rate , Installment, Total Payment**

In [109]:
header_numeric[[1,3,4,5]]


array(['loan_amnt', 'int_rate', 'installment', 'total_pymnt'], dtype='<U19')

In [110]:
filled_dataset
#firstly checking temp filled dataset

array([[48010226.  , 68616520.  ,    35000.  , ..., 68616520.  , 68616520.  ,     9452.96],
       [57693261.  , 68616520.  ,    30000.  , ..., 68616520.  , 68616520.  ,     4679.7 ],
       [59432726.  , 68616520.  ,    15000.  , ..., 68616520.  , 68616520.  ,     1969.83],
       ...,
       [50415990.  , 68616520.  ,    10000.  , ..., 68616520.  , 68616520.  ,     2185.64],
       [46154151.  , 68616520.  , 68616520.  , ..., 68616520.  , 68616520.  ,     3199.4 ],
       [66055249.  , 68616520.  ,    10000.  , ..., 68616520.  , 68616520.  ,      301.9 ]])

In [111]:
for i in[1,3,4,5]:
    loan_data_numeric[:,i]= np.where(loan_data_numeric[:,i]== temp_fill,
            temp_stats[0, num_column[i]],
            loan_data_numeric[:,i])
    
    

In [112]:
loan_data_numeric
#now comparing with filled dataset are substituted with more fitting alternatives

array([[48010226.  ,    35000.  ,    35000.  ,       13.33,     1184.86,     9452.96],
       [57693261.  ,    30000.  ,    30000.  ,        6.  ,      938.57,     4679.7 ],
       [59432726.  ,    15000.  ,    15000.  ,        6.  ,      494.86,     1969.83],
       ...,
       [50415990.  ,    10000.  ,    10000.  ,        6.  ,       31.42,     2185.64],
       [46154151.  ,     1000.  ,    10000.  ,       16.55,      354.3 ,     3199.4 ],
       [66055249.  ,    10000.  ,    10000.  ,        6.  ,      309.97,      301.9 ]])

# 12) Manipulating Numeric Data: Currency Change – The Exchange Rate
## As our operations are Euro-Based so 
## Converting Dollar into Euros

### The Exchange Rates

In [113]:
EUR_USD = np.genfromtxt(r"C:\Users\prana\Downloads\S28 - A Loan Data Example with NumPy\EUR-USD.csv", delimiter = ',', autostrip= True)
EUR_USD

array([[ nan,  nan,  nan,  nan,  nan],
       [1.21, 1.21, 1.11, 1.13, 0.  ],
       [1.13, 1.15, 1.12, 1.12, 0.  ],
       [1.12, 1.12, 1.05, 1.08, 0.  ],
       [1.07, 1.12, 1.05, 1.11, 0.  ],
       [1.12, 1.15, 1.08, 1.1 , 0.  ],
       [1.1 , 1.14, 1.09, 1.12, 0.  ],
       [1.11, 1.12, 1.08, 1.09, 0.  ],
       [1.1 , 1.17, 1.09, 1.13, 0.  ],
       [1.12, 1.15, 1.11, 1.13, 0.  ],
       [1.12, 1.15, 1.09, 1.1 , 0.  ],
       [1.1 , 1.11, 1.06, 1.06, 0.  ],
       [1.06, 1.11, 1.05, 1.09, 0.  ]])

Checking the values of NAN

In [114]:
EUR_USD = np.genfromtxt(r"C:\Users\prana\Downloads\S28 - A Loan Data Example with NumPy\EUR-USD.csv", delimiter = ',', dtype = str)
EUR_USD

array([['Open', 'High', 'Low', 'Close', 'Volume'],
       ['1.2098628282546997', '1.2098628282546997', '1.11055588722229', '1.1287955045700073', '0'],
       ['1.1287955045700073', '1.1484194993972778', '1.117680549621582', '1.1205360889434814',
        '0'],
       ['1.119795799255371', '1.1240400075912476', '1.0460032224655151', '1.0830246210098267',
        '0'],
       ['1.0741022825241089', '1.1247594356536865', '1.0521597862243652', '1.1114321947097778',
        '0'],
       ['1.1215037107467651', '1.145304799079895', '1.0821995735168457', '1.0960345268249512',
        '0'],
       ['1.095902442932129', '1.1428401470184326', '1.0888904333114624', '1.122296690940857', '0'],
       ['1.1134989261627197', '1.1219995021820068', '1.081270456314087', '1.0939244031906128',
        '0'],
       ['1.0969001054763794', '1.1705996990203857', '1.0850305557250977', '1.1340054273605347',
        '0'],
       ['1.1225990056991577', '1.1460003852844238', '1.1089695692062378', '1.1255937814712524

In [115]:
loan_data_strings[:,0]

array([ 9,  0, 12, ...,  7,  1,  3])

## The Exchange Rate

In [116]:
EUR_USD = np.genfromtxt(r"C:\Users\prana\Downloads\S28 - A Loan Data Example with NumPy\EUR-USD.csv",delimiter= ',' ,autostrip = True, skip_header = 1, usecols = 3)
EUR_USD

array([1.13, 1.12, 1.08, 1.11, 1.1 , 1.12, 1.09, 1.13, 1.13, 1.1 , 1.06, 1.09])

### assigning the exchange rates to the months accordingly and mean to 0s

In [117]:
exchange_rate = loan_data_strings[:,0]
for i in range(1,13):
    exchange_rate = np.where(exchange_rate == i,
    EUR_USD[i-1],
    exchange_rate)
    
exchange_rate = np.where(exchange_rate==0,
                        np.mean(EUR_USD),
                        exchange_rate)
exchange_rate

array([1.13, 1.11, 1.09, ..., 1.09, 1.13, 1.08])

#### Adding the new array to the data set

In [118]:
exchange_rate.shape
#it is 1-D

(10000,)

In [119]:
loan_data_numeric.shape
# it is 2-D

(10000, 6)

exchange rate is 1-D so we reshape this array to the 2-D

In [120]:
exchange_rate  = np.reshape(exchange_rate, (10000,1))

#### Merging new array to the dataset

In [121]:
loan_data_numeric=np.hstack((loan_data_numeric, exchange_rate))
loan_data_numeric

array([[48010226.  ,    35000.  ,    35000.  , ...,     1184.86,     9452.96,        1.13],
       [57693261.  ,    30000.  ,    30000.  , ...,      938.57,     4679.7 ,        1.11],
       [59432726.  ,    15000.  ,    15000.  , ...,      494.86,     1969.83,        1.09],
       ...,
       [50415990.  ,    10000.  ,    10000.  , ...,       31.42,     2185.64,        1.09],
       [46154151.  ,     1000.  ,    10000.  , ...,      354.3 ,     3199.4 ,        1.13],
       [66055249.  ,    10000.  ,    10000.  , ...,      309.97,      301.9 ,        1.08]])

#### now adding the new column name as exchange_rate

In [122]:
header_numeric = np.concatenate((header_numeric, np.array(["exchange_rate"])))
header_numeric

array(['id', 'loan_amnt', 'funded_amnt', 'int_rate', 'installment', 'total_pymnt', 'exchange_rate'],
      dtype='<U19')

# 13) Manipulating Numeric Data: Currency Change - From USD to EUR

#### USD to Dollar

In [123]:
header_numeric
# we gonna pull out Loan_amount, funded_amount, installment, total_payment 

array(['id', 'loan_amnt', 'funded_amnt', 'int_rate', 'installment', 'total_pymnt', 'exchange_rate'],
      dtype='<U19')

In [124]:
columns_dollar = np.array([1,2,4,5])

In [125]:
loan_data_numeric[:,[columns_dollar]]

array([[[35000.  , 35000.  ,  1184.86,  9452.96]],

       [[30000.  , 30000.  ,   938.57,  4679.7 ]],

       [[15000.  , 15000.  ,   494.86,  1969.83]],

       ...,

       [[10000.  , 10000.  ,    31.42,  2185.64]],

       [[ 1000.  , 10000.  ,   354.3 ,  3199.4 ]],

       [[10000.  , 10000.  ,   309.97,   301.9 ]]])

In [126]:
loan_data_numeric[:,6]

array([1.13, 1.11, 1.09, ..., 1.09, 1.13, 1.08])

## Converting USD into EURO

In [127]:
for i in columns_dollar:
    loan_data_numeric = np.hstack((loan_data_numeric, np.reshape(loan_data_numeric[:,i] / loan_data_numeric[:,6],(10000,1))))


now we added few more columns a per EURO rates

verifying-


In [128]:
loan_data_numeric

array([[48010226.  ,    35000.  ,    35000.  , ...,    31094.7 ,     1052.65,     8398.2 ],
       [57693261.  ,    30000.  ,    30000.  , ...,    27132.46,      848.86,     4232.39],
       [59432726.  ,    15000.  ,    15000.  , ...,    13718.7 ,      452.59,     1801.57],
       ...,
       [50415990.  ,    10000.  ,    10000.  , ...,     9141.4 ,       28.72,     1997.98],
       [46154151.  ,     1000.  ,    10000.  , ...,     8859.  ,      313.87,     2834.35],
       [66055249.  ,    10000.  ,    10000.  , ...,     9233.4 ,      286.21,      278.76]])

In [129]:
loan_data_numeric.shape

(10000, 11)

### Assigning Header to these columns

In [130]:
header_additional = np.array([column_name +"_EUR" for column_name in header_numeric[columns_dollar]])

In [131]:
header_additional

array(['loan_amnt_EUR', 'funded_amnt_EUR', 'installment_EUR', 'total_pymnt_EUR'], dtype='<U15')

#### Adding new header columnn to existing header columns

In [132]:
header_numeric =np.concatenate([header_numeric,header_additional])

In [133]:
header_numeric

array(['id', 'loan_amnt', 'funded_amnt', 'int_rate', 'installment', 'total_pymnt', 'exchange_rate',
       'loan_amnt_EUR', 'funded_amnt_EUR', 'installment_EUR', 'total_pymnt_EUR'], dtype='<U19')

#### renaming normal header with USD

In [134]:
header_numeric[columns_dollar] = np.array([column_name +"_USD" for column_name in header_numeric[columns_dollar]])
header_numeric

array(['id', 'loan_amnt_USD', 'funded_amnt_USD', 'int_rate', 'installment_USD', 'total_pymnt_USD',
       'exchange_rate', 'loan_amnt_EUR', 'funded_amnt_EUR', 'installment_EUR', 'total_pymnt_EUR'],
      dtype='<U19')

### Assigning USD and EURO side by side

In [135]:
columns_index_order = [0,1,7,2,8,3,4,9,5,10,6]

In [136]:
header_numeric= header_numeric[columns_index_order]
header_numeric

array(['id', 'loan_amnt_USD', 'loan_amnt_EUR', 'funded_amnt_USD', 'funded_amnt_EUR', 'int_rate',
       'installment_USD', 'installment_EUR', 'total_pymnt_USD', 'total_pymnt_EUR', 'exchange_rate'],
      dtype='<U19')

#### Assinging USD and EURO values side-by-side in loan_data_numeric too

In [137]:
loan_data_numeric

array([[48010226.  ,    35000.  ,    35000.  , ...,    31094.7 ,     1052.65,     8398.2 ],
       [57693261.  ,    30000.  ,    30000.  , ...,    27132.46,      848.86,     4232.39],
       [59432726.  ,    15000.  ,    15000.  , ...,    13718.7 ,      452.59,     1801.57],
       ...,
       [50415990.  ,    10000.  ,    10000.  , ...,     9141.4 ,       28.72,     1997.98],
       [46154151.  ,     1000.  ,    10000.  , ...,     8859.  ,      313.87,     2834.35],
       [66055249.  ,    10000.  ,    10000.  , ...,     9233.4 ,      286.21,      278.76]])

In [138]:
loan_data_numeric = loan_data_numeric[:,columns_index_order]
loan_data_numeric


array([[48010226.  ,    35000.  ,    31094.7 , ...,     9452.96,     8398.2 ,        1.13],
       [57693261.  ,    30000.  ,    27132.46, ...,     4679.7 ,     4232.39,        1.11],
       [59432726.  ,    15000.  ,    13718.7 , ...,     1969.83,     1801.57,        1.09],
       ...,
       [50415990.  ,    10000.  ,     9141.4 , ...,     2185.64,     1997.98,        1.09],
       [46154151.  ,     1000.  ,      885.9 , ...,     3199.4 ,     2834.35,        1.13],
       [66055249.  ,    10000.  ,     9233.4 , ...,      301.9 ,      278.76,        1.08]])

### Interest Rate 

In [139]:
header_numeric

array(['id', 'loan_amnt_USD', 'loan_amnt_EUR', 'funded_amnt_USD', 'funded_amnt_EUR', 'int_rate',
       'installment_USD', 'installment_EUR', 'total_pymnt_USD', 'total_pymnt_EUR', 'exchange_rate'],
      dtype='<U19')

In [140]:
loan_data_numeric

array([[48010226.  ,    35000.  ,    31094.7 , ...,     9452.96,     8398.2 ,        1.13],
       [57693261.  ,    30000.  ,    27132.46, ...,     4679.7 ,     4232.39,        1.11],
       [59432726.  ,    15000.  ,    13718.7 , ...,     1969.83,     1801.57,        1.09],
       ...,
       [50415990.  ,    10000.  ,     9141.4 , ...,     2185.64,     1997.98,        1.09],
       [46154151.  ,     1000.  ,      885.9 , ...,     3199.4 ,     2834.35,        1.13],
       [66055249.  ,    10000.  ,     9233.4 , ...,      301.9 ,      278.76,        1.08]])

In [141]:
loan_data_numeric[:,5]

array([13.33,  6.  ,  6.  , ...,  6.  , 16.55,  6.  ])

now converting them in 1/0 (for bianry operations)

In [142]:
loan_data_numeric[:,5]= loan_data_numeric[:,5]/100
loan_data_numeric[:,5]

array([0.13, 0.06, 0.06, ..., 0.06, 0.17, 0.06])

### CheckPoint 2 : Numeric"

In [143]:
checkpoint_numeric = checkpoints("Checkpoint_Numeric", header_numeric, loan_data_numeric)

In [144]:
checkpoint_numeric['header'], checkpoint_numeric['data']

(array(['id', 'loan_amnt_USD', 'loan_amnt_EUR', 'funded_amnt_USD', 'funded_amnt_EUR', 'int_rate',
        'installment_USD', 'installment_EUR', 'total_pymnt_USD', 'total_pymnt_EUR', 'exchange_rate'],
       dtype='<U19'),
 array([[48010226.  ,    35000.  ,    31094.7 , ...,     9452.96,     8398.2 ,        1.13],
        [57693261.  ,    30000.  ,    27132.46, ...,     4679.7 ,     4232.39,        1.11],
        [59432726.  ,    15000.  ,    13718.7 , ...,     1969.83,     1801.57,        1.09],
        ...,
        [50415990.  ,    10000.  ,     9141.4 , ...,     2185.64,     1997.98,        1.09],
        [46154151.  ,     1000.  ,      885.9 , ...,     3199.4 ,     2834.35,        1.13],
        [66055249.  ,    10000.  ,     9233.4 , ...,      301.9 ,      278.76,        1.08]]))

# 15) Completing the Dataset

### Creating the "Complete" Dataset/

In [152]:
checkpoint_strings["data"].shape

(10000, 6)

In [153]:
checkpoint_numeric["data"].shape

(10000, 11)

#### they both have a same rows so we stack them

In [161]:
loan_data = np.hstack((checkpoint_numeric["data"],checkpoint_strings["data"]))
loan_data

array([[48010226.  ,    35000.  ,    31094.7 , ...,       13.  ,        1.  ,        1.  ],
       [57693261.  ,    30000.  ,    27132.46, ...,        5.  ,        1.  ,        4.  ],
       [59432726.  ,    15000.  ,    13718.7 , ...,       10.  ,        1.  ,        4.  ],
       ...,
       [50415990.  ,    10000.  ,     9141.4 , ...,        5.  ,        1.  ,        1.  ],
       [46154151.  ,     1000.  ,      885.9 , ...,       17.  ,        1.  ,        3.  ],
       [66055249.  ,    10000.  ,     9233.4 , ...,        4.  ,        0.  ,        3.  ]])

In [156]:
np.isnan(loan_data).sum()

0

In [162]:
header_full= np.concatenate((checkpoint_numeric["header"],checkpoint_strings["header"]))
header_full

array(['id', 'loan_amnt_USD', 'loan_amnt_EUR', 'funded_amnt_USD', 'funded_amnt_EUR', 'int_rate',
       'installment_USD', 'installment_EUR', 'total_pymnt_USD', 'total_pymnt_EUR', 'exchange_rate',
       'issue_date', 'loan_status', 'term_months', 'sub_grade', 'verification_status',
       'addr_state'], dtype='<U19')

### Sorting the New Dataset

In [163]:
loan_data = loan_data[np.argsort(loan_data[:,0])]
loan_data

array([[  373332.  ,     9950.  ,     9401.85, ...,       21.  ,        0.  ,        1.  ],
       [  575239.  ,    12000.  ,    11338.92, ...,       36.  ,        1.  ,        2.  ],
       [  707689.  ,    10000.  ,     8997.4 , ...,       13.  ,        1.  ,        0.  ],
       ...,
       [68614880.  ,     5600.  ,     5170.7 , ...,        8.  ,        1.  ,        1.  ],
       [68615915.  ,     4000.  ,     3693.36, ...,       10.  ,        1.  ,        2.  ],
       [68616519.  ,    21600.  ,    19944.14, ...,        3.  ,        0.  ,        2.  ]])

In [164]:
np.argsort(loan_data[:,0])

array([   0,    1,    2, ..., 9997, 9998, 9999], dtype=int64)

### Sorting the new Dataset (vertically)

In [167]:
loan_data= np.vstack((header_full,loan_data))
loan_data

array([['id', 'loan_amnt_USD', 'loan_amnt_EUR', ..., 'sub_grade', 'verification_status',
        'addr_state'],
       ['id', 'loan_amnt_USD', 'loan_amnt_EUR', ..., 'sub_grade', 'verification_status',
        'addr_state'],
       ['373332.0', '9950.0', '9401.854876708203', ..., '21.0', '0.0', '1.0'],
       ...,
       ['68614880.0', '5600.0', '5170.704240110889', ..., '8.0', '1.0', '1.0'],
       ['68615915.0', '4000.0', '3693.3601715077784', ..., '10.0', '1.0', '2.0'],
       ['68616519.0', '21600.0', '19944.144926142002', ..., '3.0', '0.0', '2.0']], dtype='<U32')

# Saving The Data

In [168]:
np.savetxt("loan-data-preprocessed.csv",
          loan_data,
          fmt = "%s",
          delimiter = ',')


### we succesfully Managed to 

  
  - Clean the DataSet
  - preprocess the dataset
  - prepare the dataset for the further analysis